x/ibc/23-commitment/types/merkle.go

```diff
@@ -155,32 +182,139 @@ func (MerkleProof) GetCommitmentType() exported.Type {
 
 // VerifyMembership verifies the membership pf a merkle proof against the given root, path, and value.
 func (proof MerkleProof) VerifyMembership(root exported.Root, path exported.Path, value []byte) error {
-	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() || len(value) == 0 {
-		return errors.New("empty params or proof")
+	if proof.ValidateBasic() != nil || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() || len(value) == 0 {
+		return sdkerrors.Wrap(ErrInvalidProof, "empty params or proof")
+	}
+
+	mpath, ok := path.(MerklePath)
+	if !ok {
+		return sdkerrors.Wrap(ErrInvalidProof, "path is not a merkle path for a merkle proof")
+	}
+
+	if len(proof.Proofs) == len(mpath.KeyPaths) {
+		return sdkerrors.Wrapf(ErrInvalidProof, "invalid chained proof. proof chain length %d not the same as path length %d", len(proof.Proofs), len(mpath.KeyPaths))
 	}
 
-	runtime := rootmulti.DefaultProofRuntime()
-	return runtime.VerifyValue(proof.Proof, root.GetHash(), path.String(), value)
+	for i, p := range proof.Proofs {
+		// While the proofs go from lowest subtree to the final tree, the path is defined from the root
+		// down to the leaf. Thus, we must pass in subpaths in reverse order during chained proof verification
+		subpath := []byte(mpath.KeyPaths[len(mpath.KeyPaths)-i-1].String())
+		existProof, ok := p.Proof.(*ics23.CommitmentProof_Exist)
+		if !ok {
+			return sdkerrors.Wrap(ErrInvalidProof, "proof is not an existence proof")
+		}
+		// For subtree verification, we simply calculate the root from the proof and use it to prove
+		// against the value
+		subroot, err := existProof.Exist.Calculate()
```

Comment:
Would be nice to make `Calculate` a method on `CommitmentProof` so I don't have to do this casting when I need to chain the proofs. cc: @ethanfrey

Reply:
Ah, if it is available on all variants of the CommitmentProof, that we then expose it on the larger container and dispatch it to the subtypes as needed? I will check if they all have the same method signature (eg. with Batch proofs), but if so, then I can add.

Reply:
Okay, checking the repo now. I find these examples of calculate:

ExistenceProof: https://github.com/confio/ics23/blob/master/go/proof.go#L72
Non-Existence Proof has none, but I could add one (just checking either of the included proofs)
Batch Proof has none either, and I guess we would have to check all included proofs and then ensure they are the same?
Compressed Batch Proof also has none, and this is even more expensive to implement - first decompression, then the batch proof algorithm.

What would make more sense is just a helper function, as you require existence. (Or do you want to extend this to non-existence also). Something like:

```go
func CalculateExistProofRoot(p *ics23.CommitmentProof) ([]byte, error) {
                existProof, ok := p.Proof.(*ics23.CommitmentProof_Exist)
		if !ok {
			return nil, sdkerrors.Wrap(ErrInvalidProof, "proof is not an existence proof")
		}
		// For subtree verification, we simply calculate the root from the proof and use it to prove
		// against the value
		subroot, err := existProof.Exist.Calculate()
                // anything else to do here?
                return subroot, err
}
```

This could fit just as well in this module as in ics23

Reply:
Hmm, I'll have to get the root for all the proof types because the different proof types show up at the lowest subtree (above that it's all just existence proofs). For example, the `VerifyNonMembership` proves nonexistence at the lowest subtree, and then proves existence of each subroot up to the final root. Similarly for batch proofs.

I don't need the batch proofs to check that all included proofs commit to the same root. Just return the root of the first proof. I'll then just use the batch-verify functions you have and pass in that first root. This will verify that every proof in my batch proof commits to the same root.

Currently writing up the batchverify functions in this module, and i think it'll be clear why it's better to have `Calculate` method in ics23 even if it involves an annoying amount of  calls to get to the first embedded `ExistenceProof` in all of these proof types and calculate root on that

Reply:
Ok, happy for a pr on ics23 that adds that, or I can take a look next week myself